### PR TITLE
Dataset from modin

### PR DESCRIPTION
**Pull request by wgifford**
## Why are these changes needed?

Add support for ``from_modin()`` and ``to_modin()`` in the new `ray.data.dataset` class.

## Related issue number

#16851

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by devin-petersohn — COMMENTED** [review]
Looks right to me! @wuisawesome I left a comment for you.

File: python/requirements/data_processing/requirements.txt [hunk 1]
```diff
@@ -4,3 +4,5 @@ aioboto3==8.3.0
 flask_cors
 moto[s3]
 s3fs
+modin==0.8.3;  python_version < '3.7'
```
Comment (wuisawesome) [inline python/requirements/data_processing/requirements.txt:7]: Should this be

```suggestion
modin>=0.8.3;  python_version < '3.7'
```

Comment (wgifford) [reply python/requirements/data_processing/requirements.txt:7]: Yeah -- I think `>=` is better.

**Comment by wgifford** [comment]
@wuisawesome @devin-petersohn 
The above now includes ``to_modin/from_modin`` and it looks like tests have passed after the last update. Please let me know if there is anything else I should do.

File: python/ray/data/read_api.py [hunk 2]
```diff
@@ -453,7 +453,10 @@ def from_modin(df: "modin.DataFrame", *,
     Returns:
         Dataset holding Arrow records read from the dataframe.
     """
-    raise NotImplementedError  # P1
+    from modin.distributed.dataframe.pandas.partitions import unwrap_partitions
+
+    parts = unwrap_partitions(df, axis=0)
+    return from_pandas(parts, parallelism=parallelism)
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/read_api.py:459]: The parallelism is just the number of blocks (parallelism is preserved), so it doesn't make sense to change it here.

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -1289,12 +1289,26 @@ def to_mars(self) -> "mars.DataFrame":
     def to_modin(self) -> "modin.DataFrame":
         """Convert this dataset into a Modin dataframe.
 
+        This works by first converting this dataset into a distributed set of
+        Pandas dataframes (using ``.to_pandas()``). Please see caveats there.
+        Then the individual datarames are used to create the modin DataFrame
```
Comment (ericl) [inline python/ray/data/dataset.py:1294]:
```suggestion
        Then the individual dataframes are used to create the modin DataFrame
```

**Review by ericl — COMMENTED** [review]
Can you also update dataset.rst to have :heavy_check_mark: for from_modin / to_modin in the "Datasource Compatibility Matrices" section?

**Review by ericl — COMMENTED** [review]
Can you also update dataset.rst to have :heavy_check_mark: for from_modin / to_modin in the "Datasource Compatibility Matrices" section?

**Review by ericl — COMMENTED** [review]
Can you also update dataset.rst to have :heavy_check_mark: for from_modin / to_modin in the "Datasource Compatibility Matrices" section?

**Comment by wuisawesome** [comment]
Hey @wgifford it looks like the unit test you added is failing. https://buildkite.com/ray-project/ray-builders-pr/builds/12568#72d964a8-114c-423d-a6a5-e47d44865d09/429-1362

**Review by ericl — APPROVED** [review]

**Comment by wuisawesome** [comment]
@wgifford thanks for this contribution!

**Comment by CansuCandan** [comment]
+++++

**Pull request merged**